### DISCUSSION TITLE
For HTTP/2 after .NET 5, How does `HttpClient` handle the pending request when connection is disconnected?

### DISCUSSION THREAD

**Discussion by GF-Huang** [dotnet/runtime#70738] [Q&A] [answered]
We know that HTTP/2 has multiplexing properties. So usually only one TCP connection is established with the same server.

By reading the source code, I roughly understand that `HttpClient` internally uses `Http2Connection` to send requests. `Http2Connection` uses `Channel<T>` internally to implement the producer-consumer mode to send each request, which ensuring thread safety.

I got lost while reading due to the code complexity of `Http2Connection`. My question is: 
- When there are still multiple requests unsent in `Channel<T>`, the TCP connection (equivalents such as `NetworkStream` or `Socket`) inside `Http2Connection` is disconnected for various reasons, how are the remaining requests handled? 
- For the callers of these requests, will they get an `HttpRequestException` at the same time? 
- Or will `Http2Connection[Pool]` internally re-establish the connection and send the pending/remaining requests in the `Channel<T>`?

**Comment by huoyaoyuan** [comment]
@dotnet/ncl

**Comment by ManickaP** [comment, marked as answer]
TLDR: the connection gets reestablished automatically

There are multiple things to take into consideration:
- H/2 has a mechanism for graceful connection close from the server side: https://www.rfc-editor.org/rfc/rfc9113.html#GOAWAY. Based on that we finish certain requests and retry the remaining ones on a new connection.
- Requests do not get associated with a particular connection unless we know that the connection has a capacity for it, you can look into [GetHttp2ConnectionAsync](https://source.dot.net/System.Net.Http/R/4aef20bf39a13f39.html). So the requests associated with a connection are in flight.
- We do not retry requests in flight unless we get https://www.rfc-editor.org/rfc/rfc9113.html#name-rst_stream telling us we can, so if the connection fails, requests in flight will fail.

Hope this answers your question.
